snippets/spring-data/query.json

```diff
@@ -0,0 +1,32 @@
+    },
+    "SQL++ Query With Common Options": {
+        "prefix": ["@cbsdqopts"],
+        "body": ["public static N1QLQuery buildQueryOpts(N1QLExpression expression, JsonValue queryPlaceholderValues,","QueryScanConsistency scanConsistency, String clientContextId, boolean readonly, boolean adhoc, MutationState consistentWith, int maxParallelism, boolean metrics, int pipelineBatch, int pipelineCap, QueryProfile profile, Duration scanWait, int scanCap, JsonSerializer serializer) {","","    QueryOptions opts = QueryOptions.queryOptions();","    opts.clientContextId(clientContextId);","    opts.readonly(readonly);","    opts.adhoc(adhoc);","    opts.consistentWith(consistentWith);","    opts.maxParallelism(maxParallelism);","    opts.metrics(metrics);","    opts.pipelineBatch(pipelineBatch);","    opts.pipelineCap(pipelineCap);","    opts.profile(profile);","    opts.scanWait(scanWait);","    opts.scanCap(scanCap);","    opts.scanConsistency(scanConsistency);","    opts.serializer(serializer);","","","    // put the placeholders in the options","    if (queryPlaceholderValues instanceof JsonObject && !((JsonObject) queryPlaceholderValues).isEmpty()) {","        opts.parameters((JsonObject) queryPlaceholderValues);","    } else if (queryPlaceholderValues instanceof JsonArray && !((JsonArray) queryPlaceholderValues).isEmpty()) {","        opts.parameters((JsonArray) queryPlaceholderValues);","    }","","    return new N1QLQuery(expression, opts);","}","","","N1QLQuery query = buildQueryOpts(${1:expression}, ${2:queryPlaceholderValues}, ${3:scanConsistency}, ${4:clientContextId}, ${5:readonly}, ${6:adhoc}, ${7:consistentWith}, ${8:maxParallelism}, ${9:metrics}, ${10:pipelineBatch}, ${11:pipelineCap}, ${12:profile}, ${13:scanWait}, ${14:scanCap}, ${15:serializer});","${16:CouchbaseTemplate}.findByQuery(${17:Model}.class).matching(query);"],
+        "description": "Use SQL++ to Query the Couchbase Cluster, this has all options available."
+    },
+    "SQL++ Query With Parameters": {
+        "prefix": ["@cbsdqparam"],
+        "body": ["public static N1QLQuery buildQueryParams(N1QLExpression expression, JsonValue queryPlaceholderValues,","QueryScanConsistency scanConsistency) {","    QueryOptions opts = QueryOptions.queryOptions().scanConsistency(scanConsistency);","","    // put the placeholders in the options","    if (queryPlaceholderValues instanceof JsonObject && !((JsonObject) queryPlaceholderValues).isEmpty()) {","        opts.parameters((JsonObject) queryPlaceholderValues);","    } else if (queryPlaceholderValues instanceof JsonArray && !((JsonArray) queryPlaceholderValues).isEmpty()) {","        opts.parameters((JsonArray) queryPlaceholderValues);","    }","","    return new N1QLQuery(expression, opts);","}","","N1QLQuery query = buildQueryParams(${1:expression}, ${2:queryPlaceholderValues}, ${3:scanConsistency});","${4:CouchbaseTemplate}.findByQuery(${5:Model}.class).matching(query);"],
```

Comment:
I'm not sure it needs this complexity, it's somewhat obfuscating a simple thing the user will need to do often.. Maybe just two snippets one showing JsonArray parameters and the other JsonObject?